src/analysis/processing/qgsalgorithmserviceareafromlayer.cpp

```diff
@@ -0,0 +1,269 @@
+  if ( !startPoints )
+    throw QgsProcessingException( invalidSourceError( parameters, QStringLiteral( "START_POINTS" ) ) );
+
+  double travelCost = parameterAsDouble( parameters, QStringLiteral( "TRAVEL_COST" ), context );
+
+  bool includeBounds = true;  // default to true to maintain 3.0 API
+  if ( parameters.contains( QStringLiteral( "INCLUDE_BOUNDS" ) ) )
+  {
+    includeBounds = parameterAsBool( parameters, QStringLiteral( "INCLUDE_BOUNDS" ), context );
+  }
+
+  QVector< QgsPointXY > points;
+  QHash< int, QgsAttributes > sourceAttributes;
+  loadPoints( startPoints.get(), points, sourceAttributes, context, feedback );
+
+  feedback->pushInfo( QObject::tr( "Building graph…" ) );
+  QVector< QgsPointXY > snappedPoints;
+  mDirector->makeGraph( mBuilder.get(), points, snappedPoints, feedback );
+
+  feedback->pushInfo( QObject::tr( "Calculating service areas…" ) );
+  QgsGraph *graph = mBuilder->graph();
+
+  QgsFields fields = startPoints->fields();
+  fields.append( QgsField( QStringLiteral( "type" ), QVariant::String ) );
+  fields.append( QgsField( QStringLiteral( "start" ), QVariant::String ) );
+
+  QString pointsSinkId;
+  std::unique_ptr< QgsFeatureSink > pointsSink( parameterAsSink( parameters, QStringLiteral( "OUTPUT" ), context, pointsSinkId, fields,
+      QgsWkbTypes::MultiPoint, mNetwork->sourceCrs() ) );
+
+  QString linesSinkId;
+  std::unique_ptr< QgsFeatureSink > linesSink( parameterAsSink( parameters, QStringLiteral( "OUTPUT_LINES" ), context, linesSinkId, fields,
+      QgsWkbTypes::MultiLineString, mNetwork->sourceCrs() ) );
+
+  int idxStart;
+  QString origPoint;
+  QVector< int > tree;
+  QVector< double > costs;
+
+  int inboundEdgeIndex;
+  double startVertexCost, endVertexCost;
+  QgsPointXY startPoint, endPoint;
+  QgsGraphEdge edge;
+
+  QgsFeature feat;
+  QgsAttributes attributes;
+
+  int step =  snappedPoints.size() > 0 ? 100.0 / snappedPoints.size() : 1;
+  for ( int i = 0; i < snappedPoints.size(); i++ )
+  {
+    if ( feedback->isCanceled() )
+    {
+      break;
+    }
+
+    idxStart = graph->findVertex( snappedPoints.at( i ) );
+    origPoint = points.at( i ).toString();
+
+    QgsGraphAnalyzer::dijkstra( graph, idxStart, 0, &tree, &costs );
+
+    QgsMultiPointXY areaPoints;
+    QgsMultiPolylineXY lines;
+    QSet< int > vertices;
+
+    for ( int j = 0; j < costs.size(); j++ )
+    {
+      inboundEdgeIndex = tree.at( j );
+
+      if ( inboundEdgeIndex == -1 and j != idxStart )
+      {
+        // unreachable vertex
+        continue;
+      }
+
+      startVertexCost = costs.at( j );
+      if ( startVertexCost > travelCost )
+      {
+        // vertex is too expensive, discard
+        continue;
+      }
+
+      vertices.insert( j );
+      startPoint = graph->vertex( j ).point();
+
+      // find all edges coming from this vertex
+      for ( int edgeId : graph->vertex( j ).outgoingEdges() )
```

Comment:
```suggestion
      const QVector< int > outgoingEdges = graph->vertex( j ).outgoingEdges() ;
      for ( int edgeId : outgoingEdges )
```

Reply:
Hmm, maybe I'm wrong but we can not assign QList to QVector, can we? Replaced with `const QList<int>`, hope this is ok too.